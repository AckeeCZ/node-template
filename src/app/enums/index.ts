export enum withRelated {}
// related entities
export enum businessEnums {}
// business model enums

export default {
  // general enums
}
