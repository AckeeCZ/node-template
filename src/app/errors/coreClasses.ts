/**
 * Standard json error with possibility to add response code
 * @property {string} name Name of it
 * @extends Error
 */
export class HttpJsonError extends Error {
    /**
     * @param message Error message
     * @param status Status code for http response
     * @param errorCode Custom error code for i.e. mobile app error handling (like email is already taken)
     */
    constructor(
        public readonly status: number,
        public readonly message: string,
        public readonly errorCode: string = '',
        public readonly errorData?: object
    ) {
        super(message);
        Object.setPrototypeOf(this, new.target.prototype);
        this.name = this.constructor.name;
    }

    public toJSON() {
        return {
            message: this.message,
            status: this.status,
            errorCode: this.errorCode,
            errorData: this.errorData,
            errorClass: this.constructor.name,
            stack: this.stack,
        };
    }
}
