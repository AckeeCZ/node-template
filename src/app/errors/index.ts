import { mapValues } from 'lodash';

enum ErrorCodes {
    TEMPLATE_TEST = 'Unicorn cries',
    e0000 = "I'm a teapot",
    e2 = '',
}

// Type hacking for simplified error notation and code suggestions
export const E_CODES = (mapValues(
    ErrorCodes,
    (message: string, code: string): any => ({ message, code })
) as any) as typeof ErrorCodes;
