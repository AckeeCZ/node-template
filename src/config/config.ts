import { createLoader, safeValues, values } from 'configuru'
import { Level } from 'pino'

const loader = createLoader({
  defaultConfigPath: '.env.jsonc',
})

const configSchema = {
  logger: {
    defaultLevel: loader.custom(x => x as Level)('LOGGER_DEFAULT_LEVEL'),
    pretty: loader.bool('LOGGER_PRETTY'),
  },
  enableTests: loader.bool('ENABLE_TESTS'),
  node: {
    env: loader.string('NODE_ENV'),
  },
  server: {
    port: loader.number('SERVER_PORT'),
  },
}

export default values(configSchema)
export const safeConfig = safeValues(configSchema)
