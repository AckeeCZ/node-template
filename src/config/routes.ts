import { createRouter, defaultFinalHandler, defaultRootHandler, jsonParser } from 'ackee-node-api-core';
import httpErrorResponder from 'app/controllers/httpErrorResponder';
import { Application } from 'express';
import healthz from '../app/controllers/healthz';
import apiRoutes from '../app/routes/apiRoutes';

export default (app: Application) => {
    const router = createRouter();
    router.all('/', defaultRootHandler);
    router.use(healthz);

    router.use('/api', jsonParser(), apiRoutes);

    router.use(httpErrorResponder);
    router.use(defaultFinalHandler);
    app.use(router);
};
