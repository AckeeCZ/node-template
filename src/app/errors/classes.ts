import { E_CODES } from '.';
import * as coreErrors from './coreClasses';

type E_CODE = typeof E_CODES[keyof typeof E_CODES];

const apiErrorData = (error?: E_CODE, errorData?: object) => {
    const { message, code } = (error as any) || { message: '', code: '' };
    return [code, message, errorData] as [string, string, any];
};

/* tslint:disable:max-classes-per-file */
export class BadRequest extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(400, ...apiErrorData(error, errorData));
    }
}

export class NotAuthorized extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(403, ...apiErrorData(error, errorData));
    }
}

export class ValidationError extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(422, ...apiErrorData(error, errorData));
    }
}

export class NotAuthenticated extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(401, ...apiErrorData(error, errorData));
    }
}

export class NotFound extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(404, ...apiErrorData(error, errorData));
    }
}

export class ServerError extends coreErrors.HttpJsonError {
    constructor(error?: E_CODE, errorData?: object) {
        super(500, ...apiErrorData(error, errorData));
    }
}
