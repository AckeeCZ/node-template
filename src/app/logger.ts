import config from 'config';
import logger from 'cosmas';

export default logger(config.logger);
