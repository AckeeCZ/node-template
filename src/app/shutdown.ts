import logger from './logger'
import server from '../server'

/**
 * Handles multiple shutdown calls
 */
let shuttingDown = false

/**
 * Gracefully closes all listeners
 *
 * Process exit codes:
 *  0 - shutdown (including force shutdown)
 *  1 - caused by uncaught exception
 *  2 - error during graceful shutdown
 */
const shutdown = (event: string, force = false, exitCode = 0) => async ({
  error,
  signal,
}: {
  error?: Error
  signal?: NodeJS.Signals
}) => {
  logger.info(`Shutdown called! Event: ${event}, with exit code: ${exitCode}`)

  if (!error && !signal) return

  if (error) {
    logger.error(error)
    process.exit(exitCode)
  }
  if (signal && !['SIGINT', 'SIGTERM'].includes(signal)) {
    logger.error({ signal })
    process.exit(exitCode)
  }

  if (force) {
    process.exit(exitCode)
  }

  if (shuttingDown) {
    return
  }

  shuttingDown = true

  try {
    // Shutdown HTTP server
    await server.destroy()
    // Close connections to Databases etc.
    await Promise.all([
      //   knex.destroy(),
      //   redis.quit().catch((error: any) => {
      //     logger.error(error)
      //     redis.disconnect()
      //   }),
    ])
    logger.flush()
  } catch (error) {
    logger.error(error)
    process.exit(2)
  }
  logger.info('Server shutdown gracefully')
}

export const prepareForGracefulShutdown = () => {
  process.on('uncaughtException', error => {
    void shutdown('uncaughtException', true, 1)({ error })
  })
  process.on('SIGINT', signal => {
    void shutdown('SIGINT')({ signal })
  })
  process.on('SIGTERM', signal => {
    void shutdown('SIGTERM')({ signal })
  })
}
