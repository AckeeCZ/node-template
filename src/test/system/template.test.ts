import { E_CODES } from 'app/errors';
import { ServerError } from 'app/errors/classes';
import config from 'config';

describe('Template (System)', () => {
    test('Configuration is loaded', () => {
        expect(typeof config).toBe('object');
    });
    describe('Errors', () => {
        test('Throws', () => {
            expect(() => {
                throw new ServerError(E_CODES.TEMPLATE_TEST);
            }).toThrow(ServerError);
        });
        test('Throws with correct name', () => {
            let name = '';
            try {
                throw new ServerError(E_CODES.TEMPLATE_TEST);
            } catch (e) {
                name = e.name;
            }
            expect(name).toBe('ServerError');
        });
        test('Responds to toJson as expected', () => {
            expect(new ServerError(E_CODES.TEMPLATE_TEST, { data: true }).toJSON()).toMatchSnapshot({
                stack: expect.stringContaining('at'),
            });
        });
        test('Works without error code', () => {
            expect(new ServerError().toJSON()).toMatchSnapshot({
                stack: expect.stringContaining('at'),
            });
        });
    });
});
