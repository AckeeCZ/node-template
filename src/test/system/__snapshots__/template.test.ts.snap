// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template (System) Errors Responds to toJson as expected 1`] = `
Object {
  "errorClass": "ServerError",
  "errorCode": "Unicorn cries",
  "errorData": Object {
    "data": true,
  },
  "message": "TEMPLATE_TEST",
  "stack": StringContaining "at",
  "status": 500,
}
`;

exports[`Template (System) Errors Works without error code 1`] = `
Object {
  "errorClass": "ServerError",
  "errorCode": "",
  "errorData": undefined,
  "message": "",
  "stack": StringContaining "at",
  "status": 500,
}
`;
